### PR TITLE
Determine latest stable version of a package

### PR DESCRIPTION
**Pull request by rtaket**
The NuGet registration service provides information about each version of a package. In order to determine the latest stable version, we need to know if the package is listed and if it is prerelease. We can determine this from the registration data. Each registration item states if a package is listed (Listed==true). And since the v3 registration service uses semantic versioning for all the packages, we can determine if the package is prerelease. So we just need to walk the registration json to find the most recent package where Listed==true and version!=prerelease.

### CODE REVIEW

File: src/Ng/ElfieFromCatalogCollector.cs [hunk 1]
```diff
@@ -135,20 +135,30 @@ public ElfieFromCatalogCollector(Uri index, Storage storage, int maxThreads, Fun
...
+            RegistrationIndexPackage latestStablePackage = catalogItem.GetLatestStableVersion(this._nugetServiceUrls);
+            if (latestStablePackage == null)
+            {
+                Trace.TraceInformation("Skipping package without a released version");
+                return;
+            }
+            else if (!latestStablePackage.CatalogEntry.PackageVersion.Equals(catalogItem.PackageVersion))
             {
-                Uri packageResourceUri = await this.DownloadPackageAsync(catalogItem, registrationItem.PackageContent, cancellationToken);
+                Trace.TraceInformation("Skipping historical package");
```
Comment (jinujoseph) [inline src/Ng/ElfieFromCatalogCollector.cs:156]: what are historical package

File: src/Ng/Models/CatalogItem.cs [hunk 2]
```diff
@@ -273,6 +274,25 @@ internal RegistrationItem GetRegistrationItem(NugetServiceEndpoints nugetService
...
+
+            // Download the registration index for the package
+            using (Catalog.CollectorHttpClient client = new Catalog.CollectorHttpClient())
+            {
+                Uri registrationUrl = nugetServiceUrls.ComposeRegistrationUrl(this.PackageId);
+                string registrationIndexJson = client.GetStringAsync(registrationUrl).Result;
+                registrationIndex = RegistrationIndex.Deserialize(registrationIndexJson);
+            }
+
+            return registrationIndex.GetLatestStableVersion();
```
Comment (jinujoseph) [inline src/Ng/Models/CatalogItem.cs:273]: Do we need a null check here ?

File: src/Ng/Models/RegistrationIndex.cs [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        public RegistrationIndexPageItem[] Items
+        {
+            get;
+            private set;
+        }
+
+        /// <summary>
+        /// Gets the latest stable version of this package
+        /// </summary>
+        /// <returns>The regsitration for the latest stable version of this package.</returns>
```
Comment (jinujoseph) [inline src/Ng/Models/RegistrationIndex.cs:65]: nit pick spell check "registration"

File: src/Ng/Models/RegistrationIndexPackage.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+namespace Ng.Models
+{
+    /// <summary>
+    /// Model package registration information contained in the registartion index, 
+    /// https://api.nuget.org/v3/registration1/newtonsoft.json/index.json#items#items#items.
+    /// </summary>
+    /// <remarks>The registration index contains all the versions of a package. This type 
+    /// contains the registration information for one version of the package.</remarks>
+    public class RegistrationIndexPackage
```
Comment (jinujoseph) [inline src/Ng/Models/RegistrationIndexPackage.cs:18]: should this class inherit from RegistrationIndex ?

File: src/Ng/ElfieFromCatalogCollector.cs (hunk 1 above)
Comment (rtaket) [inline src/Ng/ElfieFromCatalogCollector.cs:156]: A released (listed) package that is not the latest version. I didn't know what to call it. Any suggestions?

File: src/Ng/Models/RegistrationIndex.cs (hunk 3 above)
Comment (rtaket) [inline src/Ng/Models/RegistrationIndex.cs:65]: Fixed the spelling of regsitration.

File: src/Ng/Models/RegistrationIndexPackage.cs (hunk 4 above)
Comment (rtaket) [inline src/Ng/Models/RegistrationIndexPackage.cs:18]: Yes, it probably should inherit RegistrationItem. They're the same. However, I meant to remove RegistrationItem from the project. I don't think it's used anymore. I'll just delete RegistrationItem.

File: src/Ng/ElfieFromCatalogCollector.cs (hunk 1 above)
Comment (jinujoseph) [inline src/Ng/ElfieFromCatalogCollector.cs:156]: ah , just a note saying the same should be good enough

Comment (rtaket) [inline src/Ng/ElfieFromCatalogCollector.cs:156]: I added a comment stating, "// The package is released, but is not the latest version."

**Comment by jinujoseph** [comment]
@srivatsn FYI

**Comment by jinujoseph** [comment]
:+1:

**Pull request merged**